src/Transports.AspNetCore/WebSockets/WebSocketHandler.cs

```diff
@@ -0,0 +1,165 @@
+        Options = options ?? throw new ArgumentNullException(nameof(options));
+        AuthorizationOptions = authorizationOptions ?? throw new ArgumentNullException(nameof(authorizationOptions));
+        _hostApplicationLifetime = hostApplicationLifetime ?? throw new ArgumentNullException(nameof(hostApplicationLifetime));
+        _authorizationService = authorizationService;
+    }
+
+    /// <inheritdoc/>
+    public virtual async Task ExecuteAsync(HttpContext httpContext, WebSocket webSocket, string subProtocol, IUserContextBuilder userContextBuilder)
+    {
+        if (httpContext == null)
+            throw new ArgumentNullException(nameof(httpContext));
+        if (webSocket == null)
+            throw new ArgumentNullException(nameof(webSocket));
+        if (subProtocol == null)
+            throw new ArgumentNullException(nameof(subProtocol));
+        if (userContextBuilder == null)
+            throw new ArgumentNullException(nameof(userContextBuilder));
+        var appStoppingToken = _hostApplicationLifetime.ApplicationStopping;
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(httpContext.RequestAborted, appStoppingToken);
+        if (cts.Token.IsCancellationRequested)
+            return;
+        try
+        {
+            using var webSocketConnection = CreateWebSocketConnection(httpContext, webSocket, cts.Token);
+            using var operationMessageReceiveStream = CreateReceiveStream(webSocketConnection, subProtocol, userContextBuilder);
+            await webSocketConnection.ExecuteAsync(operationMessageReceiveStream);
+        }
+        catch (OperationCanceledException) when (appStoppingToken.IsCancellationRequested)
+        {
+            // terminate all pending WebSockets connections when the application is in the process of stopping
+
+            // note: we are consuming OCE in this case because ASP.NET Core does not consider the task as canceled when
+            // an OCE occurs that is not due to httpContext.RequestAborted; so to prevent ASP.NET Core from considering
+            // this a "regular" exception, we consume it here
+        }
+    }
+
+    /// <summary>
+    /// Creates an <see cref="IWebSocketConnection"/>, a WebSocket message pump.
+    /// </summary>
+    protected virtual IWebSocketConnection CreateWebSocketConnection(HttpContext httpContext, WebSocket webSocket, CancellationToken cancellationToken)
+        => new WebSocketConnection(httpContext, webSocket, _serializer, Options, cancellationToken);
+
+    /// <summary>
+    /// Builds an <see cref="IOperationMessageProcessor"/> for the specified sub-protocol.
+    /// </summary>
+    protected virtual IOperationMessageProcessor CreateReceiveStream(IWebSocketConnection webSocketConnection, string subProtocol, IUserContextBuilder userContextBuilder)
+    {
+        if (subProtocol == GraphQLWs.SubscriptionServer.SubProtocol)
```

Comment:
Why did you decide to make one WebSocketHandler and not two?

Reply:
Honestly, it seems to me that it was possible to write all code without `WebSocketHandler` component at all. I think you want to provide more virtual methods here.

Reply:
Each handler is prioritized in the order of registration. For its priority, however, it may choose which of the available sub protocols to use. This is important and is why it needs to be one not two. 

The WebSocket handler is essentially a ASP.NET middleware and could be designed so. I wouldn’t do this, for a couple reasons, one of which is passing along the proper options class.  

We could roll the WebSocketHandler code directly into the main middleware and could add a few protected methods there that match the ones that exist here. 

As for adding more protected methods here, this class hardly contains any code at all as it is. What did you think I should be adding?

Reply:
See #822